File: python/experimentateur.py
import random


def echange(a: int, b: int) -> [int, int]:
    """Fonction pour échanger deux entiers"""
    temp = a
    a = b
    b = temp
    return a, b


def affiche(table: list) -> None:
    """Fonction qui affiche un tableau"""
    print(table)
    print("\n")


class Experience:
    def __init__(self, m: int, p: int) -> None:
        """
        Fonction qui cree une expérience et la stocke dans la structure passée en paramètre
        Attention : il faut que p ≤ m
        :param m: Nombre de marqueurs
        :type m: int
        :param p: Nombre de marqueurs positifs
        :type p: int
        marqueurs: Tous les marqueurs
        marqueurs_positifs: Les marqueurs positifs
        """
        self.m = m

        self.marqueurs = [i for i in range(0, m)]
        # random.shuffle uses fisher-yates algorithm
        random.shuffle(self.marqueurs)

        tmp = [i for i in range(0, m)]
        self.p = p
        self.marqueurs_positifs = []
        for i in range(0, p):
            j = random.randint(0, m - 1 - i)
            self.marqueurs_positifs.insert(i, self.marqueurs[tmp[j]])
            tmp[j] = tmp[m - i - 1]

    def free_xp(self):
        self.marqueurs.clear()
        self.marqueurs_positifs.clear()


def merge(arr: list, left_index: int, middle_index: int, right_index: int) -> None:
    """
    Algorithme de tri fusion
    Merges two subarrays of arr[].
    First subarray is arr[l..m]
    Second subarray is arr[m+1..r]
    :param arr:
    :param left_index:
    :param middle_index:
    :param right_index:
    :return:
    """
    n1 = middle_index - left_index + 1
    n2 = right_index - middle_index
    left, right = [], []
    for i in range(0, n1):
        left.insert(i, arr[left_index + i])
    for j in range(0, n2):
        right.insert(j, arr[middle_index + 1 + j])

    i, j, k = 0, 0, left_index
    while i < n1 and j < n2:
        if left[i] <= right[j]:
            arr[k] = left[i]
            i += 1
        else:
            arr[k] = right[j]
            j += 1
        k += 1

    while i < n1:
        arr[k] = left[i]
        i += 1
        k += 1

    while j < n2:
        arr[k] = right[j]
        j += 1
        k += 1


def merge_sort(arr: list, left_index: int, right_index: int) -> None:
    """
    l is for left index and r is right index of the sub-array of arr to be sorted
    :param arr:
    :param left_index:
    :param right_index:
    :return:
    """
    if left_index < right_index:
        m = left_index + (right_index - left_index) // 2

        merge_sort(arr, left_index, m)
        merge_sort(arr, m + 1, right_index)
        merge(arr, left_index, m, right_index)


def marqueurs_negatifs1(xp: Experience) -> [list]:
    """
    Fonction a completer - Stratégie 1
    :param xp:
    :return:
    """
    cpt_op = 0
    res = []
    for elt in xp.marqueurs:
        found, tmpop = rech_seq(xp.marqueurs_positifs, elt)
        cpt_op+=tmpop
        if not found:
            res.append(elt)
    return res, cpt_op


def rech_seq(array, marqueur_rech):
    op=0
    for elt in array:
        op += 1
        if elt == marqueur_rech:
            return True, op
    return False, op


def marqueurs_negatifs2(xp: Experience) -> [list]:
    """
    Fonction a completer - Stratégie 2
    :param xp:
    :return:
    """
    xpn = xp.m - xp.p
    res = []
    merge_sort(xp.marqueurs_positifs, 0, xp.p - 1)
    cpt_op = 0
    for elt in xp.marqueurs:
        trouver, tmpop = rechercher_dico(xp.marqueurs_positifs, xp.p, elt)
        cpt_op += tmpop
        if not trouver:
            res.append(elt)
        if len(res) == xpn:
            break

    return res, cpt_op


def rechercher_dico(table: list, n: int, elt: int):
    bas, op, haut = 0,0, n - 1
    while True:
        op += 1
        milieu = (bas + haut) // 2
        if elt == table[milieu]:
            return True, op
        elif table[milieu] < elt:
            bas = milieu + 1
        else:
            haut = milieu - 1
        if not bas <= haut:
            break
    return False, op


def marqueurs_negatifs3(xp: Experience) -> [list]:
    """
    Fonction a completer - Stratégie 3
    :param xp:
    :return:
    """
    nn = xp.m - xp.p
    cpt_op, debut, pos = 0, 0, 0
    res = []
    merge_sort(xp.marqueurs, 0, xp.m - 1)
    merge_sort(xp.marqueurs_positifs, 0, xp.p - 1)
    for p in xp.marqueurs_positifs:
        cpt_op += 1
        pos = position(xp.marqueurs, p)
        for i in range(debut, pos):
            cpt_op += 1
            res.append(xp.marqueurs[i])
        debut = pos + 1
        if len(res) == nn:
            break
    if len(res) < nn:
        for i in range(pos + 1, xp.m):
            cpt_op += 1
            res.append(xp.marqueurs[i])

    return res, cpt_op


def position(t, elt):
    for i in range(len(t)):
        if t[i] == elt:
            return i
    return -1


def test(p: int, m: int):
    xp = Experience(m, p)
    print("Marqueurs :\n")
    affiche(xp.marqueurs)
    print("Marqueurs positifs :\n")
    affiche(xp.marqueurs_positifs)

    print("Stratégie 1\n")
    marqueurs_negatifs, cpt1 = marqueurs_negatifs1(xp)
    print("Marqueurs négatifs :\n")
    affiche(marqueurs_negatifs)
    print("Stratégie 1 / Nombres d'opérations : ", cpt1, "\n\n")
    marqueurs_negatifs.clear()

    print("Stratégie 2\n")
    marqueurs_negatifs, cpt2 = marqueurs_negatifs2(xp)
    print("Marqueurs négatifs :\n")
    affiche(marqueurs_negatifs)
    print("Stratégie 2 / Nombres d'opérations : ", cpt2, "\n\n")
    marqueurs_negatifs.clear()

    print("Stratégie 3\n")
    marqueurs_negatifs, cpt3 = marqueurs_negatifs3(xp)
    print("Marqueurs négatifs :\n")
    affiche(marqueurs_negatifs)
    print("Stratégie 3 / Nombres d'opérations : ", cpt3, "\n\n")
    marqueurs_negatifs.clear()

    res_list = [m, p, cpt1, cpt2, cpt3]
    return res_list

File: experimentateur.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include "experimentateur.h"
#include <stdbool.h>

int rechercheDicho(int *pInt, int p, int i, int *pInt1);

int position(int *pInt, int m, int i);

// Fonction pour echanger deux entiers
void echange(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

// Fonction qui affiche un tableau de taille n
void affiche(int *t, int n) {
    int i;

    for (i = 0; i < n; ++i)
        printf("%d ", t[i]);
    printf("\n");
}

// Fonction qui cree une expérience et la stocke dans la structure passee en parametre
// Attention : il faut que p <= m
void cree_experience(EXPERIENCE *xp, int p, int m) {
    int i, j;

    xp->m = m;

    xp->marqueurs = (int *) malloc(m * sizeof(int));
    for (i = 0; i < m; ++i)
        xp->marqueurs[i] = i;

    // On melange les marqueurs dans l'ensemble des marqueurs par l'algorithme de Fisher-Yates
    for (i = m - 1; i >= 0; --i) {
        j = rand() % (i + 1);
        echange(&xp->marqueurs[i], &xp->marqueurs[j]);
    }

    // On choisit p marqueurs parmi les m qui seront les marqueurs positifs
    int *tmp = (int *) malloc(m * sizeof(int));
    for (i = 0; i < m; ++i)
        tmp[i] = i;

    xp->p = p;

    xp->marqueurs_positifs = (int *) malloc(p * sizeof(int));
    for (int i = 0; i < p; ++i) {
        j = rand() % (m - i);
        xp->marqueurs_positifs[i] = xp->marqueurs[tmp[j]];
        tmp[j] = tmp[m - i - 1];
    }

    free(tmp);
}

// Algorithme de tri fusion
// Merges two subarrays of arr[].
// First subarray is arr[l..m]
// Second subarray is arr[m+1..r]
void merge(int arr[], int l, int m, int r) {
    int i, j, k;
    int n1 = m - l + 1;
    int n2 = r - m;

    /* create temp arrays */
    int L[n1], R[n2];

    /* Copy data to temp arrays L[] and R[] */
    for (i = 0; i < n1; i++)
        L[i] = arr[l + i];
    for (j = 0; j < n2; j++)
        R[j] = arr[m + 1 + j];

    /* Merge the temp arrays back into arr[l..r]*/
    i = 0; // Initial index of first subarray
    j = 0; // Initial index of second subarray
    k = l; // Initial index of merged subarray
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    /* Copy the remaining elements of L[], if there
     are any */
    while (i < n1) {
        arr[k] = L[i];
        i++;
        k++;
    }

    /* Copy the remaining elements of R[], if there
     are any */
    while (j < n2) {
        arr[k] = R[j];
        j++;
        k++;
    }
}

/* l is for left index and r is right index of the
 sub-array of arr to be sorted */
void mergeSort(int arr[], int l, int r) {
    if (l < r) {
        // Same as (l+r)/2, but avoids overflow for
        // large l and h
        int m = l + (r - l) / 2;

        // Sort first and second halves
        mergeSort(arr, l, m);
        mergeSort(arr, m + 1, r);

        merge(arr, l, m, r);
    }
}

// Fonction qui libere la memoire dans une experience
void libere_experience(EXPERIENCE *xp) {
    free(xp->marqueurs);
    free(xp->marqueurs_positifs);
}

// Fonction a completer - Strategie 1
// Le second argument servira a compter le nombre d'utilisation de l'opérateur OP
int *marqueurs_negatifs1(EXPERIENCE *xp, int *cptOP) {
    *cptOP = 0;

    int *res = (int *) malloc((xp->m - xp->p) * sizeof(int));

    return res;
}

// Fonction a completer - Strategie 2
// Le second argument servira a compter le nombre d'utilisation de l'opérateur OP
int *marqueurs_negatifs2(EXPERIENCE *xp, int *cptOP) {
    *cptOP = 0;
    int nn=(xp->m - xp->p);
    int *res = (int *) malloc(nn * sizeof(int));
    mergeSort(xp->marqueurs_positifs, 0, xp->p-1);
    int j=-1;
    for (int i = 0; i < xp->m; i++) {
        (*cptOP)++;
        int trouver=rechercheDicho(xp->marqueurs_positifs,xp->p,xp->marqueurs[i],cptOP);
        if(trouver==-1){
            j++;
            res[j]=xp->marqueurs[i];
        }
        if (j==nn+1) break;
    }
    return res;
}
int rechercheDicho(int *t, int n, int elt, int *op) {
    int bas=0,haut=n-1,milieu;
    //printf("->elt = %d \n",elt);
    do{
        (*op)++;
        milieu=(bas+haut)/2;
        //printf("milieu = %d \n",milieu);
        if (elt==t[milieu]) return milieu;
        else if (t[milieu]<elt) bas=milieu+1;
        else haut=milieu-1;
    }while (bas<=haut);
    return -1;
}
// Fonction a completer - Strategie 3
// Le second argument servira a compter le nombre d'utilisation de l'opérateur OP
int *marqueurs_negatifs3(EXPERIENCE *xp, int *cptOP) {
    *cptOP = 0;
    int nn=(xp->m - xp->p);
    int *res = (int *) malloc((xp->m - xp->p) * sizeof(int));
    mergeSort(xp->marqueurs, 0, xp->m-1);
    mergeSort(xp->marqueurs_positifs, 0, xp->p-1);
    int debut=0,iDeN=-1;
    int pos;
    for (int p = 0; p < xp->p; ++p) {
        (*cptOP)++;
        pos=position(xp->marqueurs,xp->m,xp->marqueurs_positifs[p]);
        for (int i = debut; i < pos; ++i) {
            (*cptOP)++;
            iDeN++;
            res[iDeN]=xp->marqueurs[i];
        }
        debut=pos+1;
        if (iDeN==nn+1) break;
    }
    if (pos<xp->m-1){
        for (int i = pos+1; i <xp->m ; ++i) {
            (*cptOP)++;
            iDeN++;
            res[iDeN]=xp->marqueurs[i];
        }
    }
    return res;
}

int position(int *t, int n, int elt) {
    for (int j = 0; j < n; ++j) {
        if (t[j]==elt) return j;
    }
    return -1;
}

void test(int p, int m) {
    EXPERIENCE xp;
    int cpt, *marqueurs_negatifs;

    // Creation de l'experience
    cree_experience(&xp, p, m);

    printf("Marqueurs :\n");
    affiche(xp.marqueurs, m);
    printf("\nMarqueurs positfis :\n");
    affiche(xp.marqueurs_positifs, p);

    //marqueurs_negatifs = marqueurs_negatifs1(&xp, &cpt);
    //marqueurs_negatifs = marqueurs_negatifs2(&xp, &cpt);
    marqueurs_negatifs = marqueurs_negatifs3(&xp, &cpt);

    printf("\nStrategie: \n");
    printf("Marqueurs negatifs :\n");
    affiche(marqueurs_negatifs, xp.m - xp.p);
    printf("Strategie / nombres OP : %d\n\n", cpt);
    free(marqueurs_negatifs);

    libere_experience(&xp);
}